### PR TITLE
[prometheus-node-exporter] Patch nodeSelector on node-exporter daemonset

### PR DESCRIPTION
**Pull request by marciogmorales**
#### What this PR does / why we need it

On a heterogenous Kubernetes cluster with Windows nodes, the daemonset pods try to be scheduled on all nodes as it doesn't have nodeSelector by OS type, and Linux is the only one supported for node-exporter.

This proposed change adds the kubernetes.io/os: Linux as nodeSelector in order to avoid failed loop on pod scheduling on Windows nodes.

#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #

#### Special notes for your reviewer

I'm starting addressing the node-exporter daemonset, however the same fix will be needed on other prometheus deployments, which I'll work on once this PR is merged to the main.

#### Checklist

- [X] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [X] Chart Version bumped
- [X] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: charts/prometheus-node-exporter/templates/daemonset.yaml [hunk 1]
```diff
@@ -39,6 +39,8 @@ spec:
         {{- toYaml . | nindent 8 }}
       {{- end }}
       serviceAccountName: {{ include "prometheus-node-exporter.serviceAccountName" . }}
+      nodeSelector:
+        kubernetes.io/os: {{ .Values.nodeselector.os }}
```
Comment (zanhsieh) [inline charts/prometheus-node-exporter/templates/daemonset.yaml:43]: I would let the users picks their own generic purpose spec here:

```suggestion
      {{- with .Values.nodeselector }}
      nodeSelector:
        {{- toYaml . | nindent 8 }}
      {{- end }}
```

File: charts/prometheus-node-exporter/values.yaml [hunk 2]
```diff
@@ -434,3 +434,8 @@ verticalPodAutoscaler:
     # Specifies whether recommended updates are applied when a Pod is started and whether recommended updates
     # are applied during the life of a Pod. Possible values are "Off", "Initial", "Recreate", and "Auto".
     # updateMode: Auto
+    
+ ## NodeSelector
+ # It forces prometheus-node-exporter to be only scheduled on Linux-based worker nodes as it isn't compatible with Windows OS.
+nodeselector:
+  os: linux
```
Comment (zanhsieh) [inline charts/prometheus-node-exporter/values.yaml:441]: Set to `{}` as backward compatible.
```suggestion
nodeselector: {}
#   kubernetes.io/os: linux
```

**Review by zanhsieh — CHANGES_REQUESTED** [review]
@marciogmorales 
Thanks for the contribution. Beside the change I suggested, would you mind to bump the version in Chart.yaml so the CI can merge this PR please?

**Comment by marciogmorales** [comment]
@zanhsieh I think we will need to force the nodeSelector to Linux OS and not let it as optional value, because the Chart deployment fails on an heterogeneous cluster. Thoughts?

**Comment by zanhsieh** [comment]
> @zanhsieh I think we will need to force the nodeSelector to Linux OS and not let it as optional value, because the Chart deployment fails on an heterogeneous cluster. Thoughts?

As long as that's backward compatible, I am fine with it.

**Comment by marciogmorales** [comment]
> 

Should I re-write the code as I did in the first modification?

**Review by marciogmorales — COMMENTED** [review]
Fixed several inconsistent that parameters were referenced to wrong values. Added nodeSelector to be mapped to Linux instead of an optional value.

File: charts/prometheus-node-exporter/templates/daemonset.yaml [hunk 3]
```diff
@@ -220,18 +225,11 @@ spec:
...
         {{- toYaml . | nindent 8 }}
       {{- end }}
-      {{- with .Values.nodeSelector }}
-      nodeSelector:
-        {{- toYaml . | nindent 8 }}
-      {{- end }}
-      {{- with .Values.tolerations }}
-      tolerations:
-        {{- toYaml . | nindent 8 }}
-      {{- end }}
```
Comment (zanhsieh) [inline charts/prometheus-node-exporter/templates/daemonset.yaml:234]: @marciogmorales 
Could you put back the `tolerations` please?

**Review by zanhsieh — CHANGES_REQUESTED** [review]

**Review by marciogmorales — COMMENTED** [review]
Removing empty 278 line

File: charts/prometheus-node-exporter/templates/daemonset.yaml [hunk 4]
```diff
@@ -39,6 +39,7 @@ spec:
         {{- toYaml . | nindent 8 }}
       {{- end }}
       serviceAccountName: {{ include "prometheus-node-exporter.serviceAccountName" . }}
+      {{- with .Values.nodeselector }}
```
Comment (zanhsieh) [inline charts/prometheus-node-exporter/templates/daemonset.yaml:42]: @marciogmorales 
You miss an `{{- end }}` in line 278 if you'd like to keep `{{- with .Values.nodeselector }}`.

**Review by zanhsieh — CHANGES_REQUESTED** [review]

**Comment by marciogmorales** [comment]
Thank you @zanhsieh . Fixed the issue.

**Comment by jcpunk** [comment]
I'll confess some interest in this.

File: charts/prometheus-node-exporter/templates/daemonset.yaml [hunk 5]
```diff
@@ -228,15 +229,15 @@ spec:
       affinity:
         {{- toYaml . | nindent 8 }}
       {{- end }}
-      {{- with .Values.dnsConfig }}
+      {{- with .Values.affinity }}
```
Comment (jcpunk) [inline charts/prometheus-node-exporter/templates/daemonset.yaml:232]: These `with` blocks look wrong to me...

**Pull request merged**